File: find min rotated sorted arr.js
/*
 * array of length n sorted in ascending order is rotated between 1 and n times. For example, the array nums = [0,1,2,4,5,6,7] might become:
 * [4,5,6,7,0,1,2] if it was rotated 4 times.
 * [0,1,2,4,5,6,7] if it was rotated 7 times.
 * algorithm that runs in O(log n) time.
 * return the minimum element of this array.
 */

var findMin = function (nums) {
    let start = 0, end = nums.length - 1, mid;
    while (true) {
        mid = Math.floor((start + end) / 2);
        if (nums[mid + 1] < nums[mid]) {
            return nums[mid + 1];
        }
        else if (nums[mid] < nums[mid - 1]) {
            return nums[mid];
        }
        else if (nums[end] > nums[mid]) {
            end = mid - 1;
        }
        else {
            start = mid + 1;
        }
    }
};

console.log(findMin([4, 5, 6, 7, 0, 1, 2]));


/*
 * The main idea is, the element is said to be minimum in the rotated sorted array if the previous element to it is greater than it or there is no previous element(i.e. no rotation). We can do this using Binary search
 * Find the mid element i.e. mid = (low+high)/2
 * If the (mid+1)th element is less than mid element then return (mid+1)th element
 * If the mid element is less than (mid-1)th element then return the mid element
 * If the last element is greater than mid element then search in left half
 * If the last element is less than mid element then search in right half
 * Time Complexity: O(logn)
 * Space Complexity: O(1)
 */

// method 2

var findMin1 = function (nums) {
    let left = 0, right = nums.length - 1;
    while (left < right) {
        let mid = Math.floor((left + right) / 2);

        if (nums[mid] < nums[right]) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    return nums[left];
};

console.log(findMin1([4, 5, 6, 7, 0, 1, 2]));

File: Find_min_rotated_sorted_arr.java
/**
 * Find_min_rotated_sorted_arr
 * Suppose an array of length n sorted in ascending order is rotated between 1 and n times. For example, the array nums = [0,1,4,4,5,6,7] might become:
 * [4,5,6,7,0,1,4] if it was rotated 4 times
 * [0,1,4,4,5,6,7] if it was rotated 7 times.
 * 
 * Given the sorted rotated array nums that may contain duplicates, return the minimum element of this array.
 * 
 * https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/description/
 */
public class Find_min_rotated_sorted_arr {

    public static void main(String[] args) {
        int[] arr = { 3, 1, 3, 3, 3, 3, 3 };
        System.out.println(findMin(arr));
    }
    
    static public int findMin(int[] nums) {
        return binearySearch(nums, 0, nums.length - 1);
    }

    static public int binearySearch(int[] arr, int start, int end){
        // base condition
        if(end - start == 1){
            return Math.min(arr[end],arr[start]);
        }

        while (start < end){
            int mid = start + (end - start)/2;

            if(arr[mid] > arr[mid+1] ){
                return arr[mid+1];
            }
            else if(arr[mid] < arr[end]){
                end = mid;
            }
            // this is unconditional deciding state so, we check both half
            else if(arr[mid] == arr[end]){
                return Math.min(binearySearch(arr,mid+1,end),binearySearch(arr,start,mid));
            }
            else{
                start = mid + 1;
            }
        }

        return arr[start];

    }
}